Publish workspace

jneem opened the pull request:
<!-- homu-ignore:start -->
<!--
Thanks for submitting a pull request 🎉! Here are some tips for you:

* If this is your first contribution, read "Cargo Contribution Guide" first:
  https://doc.crates.io/contrib/
* Run `cargo fmt --all` to format your code changes.
* Small commits and pull requests are always preferable and easy to review.
* If your idea is large and needs feedback from the community, read how:
  https://doc.crates.io/contrib/process/#working-on-large-features
* Cargo takes care of compatibility. Read our design principles:
  https://doc.crates.io/contrib/design.html
* When changing help text of cargo commands, follow the steps to generate docs:
  https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages
* If your PR is not finished, set it as "draft" PR or add "WIP" in its title.
* It's ok to use the CI resources to test your PR, but please don't abuse them.

### What does this PR try to resolve?

Explain the motivation behind this change.
A clear overview along with an in-depth explanation are helpful.

You can use `Fixes #<issue number>` to associate this PR to an existing issue.

### How should we test and review this PR?

Demonstrate how you test this change and guide reviewers through your PR.
With a smooth review process, a pull request usually gets reviewed quicker.

If you don't know how to write and run your tests, please read the guide:
https://doc.crates.io/contrib/tests

### Additional information

Other information you want to mention in this PR, such as prior arts,
future extensions, an unresolved problem, or a TODO list.
-->
<!-- homu-ignore:end -->

Adds support for simultaneously publishing multiple (possibly inter-dependent) packages in a workspace, gated by the `-Zpackage-workspace` flag.

We had a couple of questions:
- Is it worth adding a new `-Zpublish-workspace` flag, or is it ok to reuse the old one?
- This breaks the `publish_namespaced` test, because we now try to parse all packages as package specs, and "foo::bar" fails to parse. This could probably be avoided by a refactor (adding a different package function that takes a list of packages instead of trying to get them from cli arguments), but I also think it's a bug in the package spec parser. Thoughts?

ehuss commented:
> Is it worth adding a new `-Zpublish-workspace` flag, or is it ok to reuse the old one?

I don't have enough context to know if `-Zpackage-workspace` was intended as a stepping-stone for publishing, or if there really is an independent use-case for packaging versus publishing.

> This breaks the `publish_namespaced` test, because we now try to parse all packages as package specs, and "foo::bar" fails to parse. This could probably be avoided by a refactor (adding a different package function that takes a list of packages instead of trying to get them from cli arguments), but I also think it's a bug in the package spec parser. Thoughts?

Can the [check here](https://github.com/rust-lang/cargo/blob/ef854d2f66df7bfcd803a6ca1cfa3b619603142c/src/cargo/util/toml/mod.rs#L2707-L2714) be moved to an earlier stage, like in `ops::registry::publish::publish` by iterating over the `just_pkgs` list and checking if any of them have the `open-namespaces` feature?

ehuss commented:
I realistically won't have time to give this a good review for at least a few days if not a week or more. I was wondering if @weihanglo or @epage would have more bandwidth in the short term? If not, that's ok, I'll try to review this when I can.

epage commented:
> This breaks the publish_namespaced test, because we now try to parse all packages as package specs, and "foo::bar" fails to parse. This could probably be avoided by a refactor (adding a different package function that takes a list of packages instead of trying to get them from cli arguments), but I also think it's a bug in the package spec parser. Thoughts?

@jneem I posted #14467

epage commented:
Could you rebase on top of #14467?

jneem commented:
Rebased, thanks!

epage commented on src/cargo/ops/cargo_package.rs:
```diff
@@ -187,7 +187,11 @@ pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Vec<Fi
         }
     }
-    let pkgs = ws.members_with_features(specs, &opts.cli_features)?;
+    let mut pkgs = ws.members_with_features(specs, &opts.cli_features)?;
+
+    // In `members_with_features_old`, it will add "current" package (determined by the cwd)
+    // So we need filter
+    pkgs.retain(|(pkg, _feats)| specs.iter().any(|spec| spec.matches(pkg.package_id())));
```
This change mirrors similar code in `cargo publish` that was added in #10677.

epage replied:
Could you add that as a note to the commit?

btw I think this and the prior commit could be their own PR?

epage commented:
> > Is it worth adding a new -Zpublish-workspace flag, or is it ok to reuse the old one?
>
> I don't have enough context to know if -Zpackage-workspace was intended as a stepping-stone for publishing, or if there really is an independent use-case for packaging versus publishing.

`-Zpackage-workspace` is a stepping stone.  In that spirit, it seems like the feature would be named `-Zpublish-workspace` but probably not worth changing at this point.

The questions relevant for this is would we want to stabilize these together?  Are there design risks for one that by separating the stabilization of them would make it easier to stabilize the other?

epage commented on tests/testsuite/publish.rs:
We should have a test case for the publish equivalent of #14356.

jneem replied:
I brought one of those tests over, but this PR's behavior is intentionally different from `cargo package`: we abort with an error as soon as any package in our list has `publish = false`. Maybe we could be more permissive, by filtering out unpublishable packages if `to_publish` is `Packages::Default`?

epage replied:
Whichever way we go, let's call it out in the PR description and in the tracking issue / stabilization report so we can make sure we make a conscious decision about this.

For myself, I think I lean towards `--workspace` implicitly excluding packages while `--package foo` should fail.  Unsure how cleanly we can implement those semantics.  However, erroring is the safer choice as we could probably change to implicit exclude later.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+                        registry
+                    ))?;
+                }
             }
```
nit: looks like switching to `infer_registry` could be broken out into a refactor commit?  Within the existing single-package publish, it should have no behavior change but its a stepping stone towards multi-package publish.

epage replied:
Thanks!  This makes it really clear that that commit is actually diagnostic fix!

(technically, it could be pulled out into its own PR then... but not going to block on that)

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
 
-    let (pkg, cli_features) = pkgs.pop().unwrap();
-
-    let mut publish_registry = match opts.reg_or_index.as_ref() {
-        Some(RegistryOrIndex::Registry(registry)) => Some(registry.clone()),
-        _ => None,
-    };
-    if let Some(ref allowed_registries) = *pkg.publish() {
-        if publish_registry.is_none() && allowed_registries.len() == 1 {
```
nit: a potential refactor that could be broken out is to switch this code into the style of `match` in the new code.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
     // This is only used to confirm that we can create a token before we build the package.
     // This causes the credential provider to be called an extra time, but keeps the same order of errors.
-    let ver = pkg.version().to_string();
-    let operation = Operation::Read;
-
-    let reg_or_index = match opts.reg_or_index.clone() {
-        Some(RegistryOrIndex::Registry(_)) | None => {
-            publish_registry.map(RegistryOrIndex::Registry)
-        }
-        val => val,
-    };
     let source_ids = super::get_source_id(opts.gctx, reg_or_index.as_ref())?;
     let mut registry = super::registry(
         opts.gctx,
         &source_ids,
         opts.token.as_ref().map(Secret::as_deref),
         reg_or_index.as_ref(),
         true,
-        Some(operation).filter(|_| !opts.dry_run),
```
nit: removing `operation` variable could be its own refactor commit

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-    // Prepare a tarball, with a non-suppressible warning if metadata
-    // is missing since this is being put online.
-    let tarball = ops::package_one(
+    // Validate all the packages before publishing any of them.
+    for (pkg, _) in &pkgs {
+        verify_dependencies(pkg, &registry, source_ids.original)?;
+    }
```
note for people reviewing this: at this time, `verify_dependencies` does not contact the registry, so the local, unpublished dependencies shouldn't be a problem

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+    let specs = Packages::from_flags(
+        false,
+        vec![],
+        just_pkgs.iter().map(|p| p.name().to_string()).collect(),
```
package names are not always sufficient specs.  I don't know what paths this code goes down but if it hits the wrong one now, or in the future, this will break in less commonly seen but important cases.

For example, `clap` has transitive dependencies on its self so I can't run `cargo check -p clap`, I have to do `cargo check -p clap@...`

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+        verify_dependencies(pkg, &registry, source_ids.original)?;
+    }
+
+    let specs = Packages::from_flags(
+        false,
+        vec![],
+        just_pkgs.iter().map(|p| p.name().to_string()).collect(),
+    )?;
+    let pkg_dep_graph = ops::cargo_package::package_with_dep_graph(
```
In general, it feels odd for us to have process the specs and then convert them back into specs and process them again.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+    while !to_confirm.is_empty() {
+        if !ready.is_empty() {
+            opts.gctx.shell().status(
+                "Uploading",
+                ready.iter().map(PackageId::to_string).sorted().join(", "),
+            )?;
+        }
+
+        for pkg_id in &ready {
+            let (pkg, (_features, tarball)) = &pkg_dep_graph.packages[pkg_id];
+
+            if !opts.dry_run {
+                let ver = pkg.version().to_string();
+
+                tarball.file().seek(SeekFrom::Start(0))?;
```
While I usually try to minimize noise, we should print as we are uploading because its important context for when an upload fails.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+                // Short does not include the registry name.
+                let short_pkg_description = format!("{} v{}", pkg.name(), pkg.version());
+                let source_description = source_ids.original.to_string();
+                ws.gctx().shell().status(
+                    "Uploaded",
+                    format!("{short_pkg_description} to {source_description}"),
+                )?;
```
At the time, the uploaded message was important context for what phase we are at with waiting.  I wonder if its just extra noise in this scenario.

epage replied:
Looking over how this is done, maybe it does make sense.  Unsure

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+        }
+        if finished.is_empty() {
+            // If nothing finished, it means we timed out while waiting for confirmation.
+            // We're going to exit, but first we need to check: if we pretend that the unconfirmed
+            // uploads are confirmed, did we finish everything?
```
What does this mean?

jneem replied:
Prior to multi-package publishing, it wasn't an error if the verification timed out: we just assume that it will appear in the registry eventually. We try to imitate the same behavior when publishing multiple packages, but we *do* raise an error if the package(s) that timed out had dependencies that we haven't uploaded yet.

To be precise, we exit with success if and only if "confirmed packages" union "uploaded but not confirmed packages" is everything.

epage replied:
Thanks.  I think the way the code is structured can be a bit confusing.  I'll try to figure out what led me to be confused and how to improve it.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+    while !to_confirm.is_empty() {
+        if !ready.is_empty() {
```
Why do we continue the loop when `ready` is empty?

jneem replied:
Say we have packages `main`, `dep1`, and `dep2`, where `main` depends on both of the `dep`s. The first time through the loop, both of the deps are ready, so we upload them both. But now suppose that when we poll, we only see `dep1` as ready. Then we'll run through the loop again even though nothing is ready for upload.

Having said that, I'm going to remove this check in favor of printing the "Uploading..." messages one-by-one for more precise context.

epage replied:
> Say we have packages main, dep1, and dep2, where main depends on both of the deps. The first time through the loop, both of the deps are ready, so we upload them both. But now suppose that when we poll, we only see dep1 as ready. Then we'll run through the loop again even though nothing is ready for upload.

That we are doubling the timeout for some packages is non-obvious and this should probably have a comment in the code.

epage replied:
The comment that was added to resolve this, while good and covering literally what I asked for, didn't fully address the concern I had. The comment was put deeper, inside of `if 0 < timeout {` but this doesn't help someone to understand why you can have `remaining` without `ready` where they are reading this non-obvious code.

jneem replied:
Ah, since I'd removed the `!ready.is_empty()` check it wasn't clear to me that the explanation was still needed. I've added one, at the point where we retrieve the ready packages.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -210,34 +258,28 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
 fn wait_for_publish(
```
Oh, we are waiting until the first set of packages become available.

We should probably rename this function to clarify that.

epage replied:
`wait_for_any_publish_confirmation`

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -247,43 +289,67 @@ fn wait_for_publish(
+    Ok(available)
+}
+
+fn poll_one_package(
```
nit: could pulling this out be its own refactor commit?

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -498,3 +564,152 @@ fn transmit(
+impl PublishOrder {
+    // Given a package dependency graph, creates a `PublishOrder` for tracking state and also
+    // a list of packages that have no dependencies (i.e. can be published immediately).
+    fn new(graph: &Graph<PackageId, ()>) -> (Self, Vec<PackageId>) {
```
API wise this feels weird that we return `ready` as well as `Self`.

Options
- We keep it.  Its all internal
- We add a `take_ready` which removes items with `weight == 0`
- likely other approaches as well

jneem replied:
The API is definitely a little weird, but it has the advantage that it requires the least state-tracking: when you mark a package as finished, it's immediately obvious (in constant time, with no extra state) which packages are ready. To have a two-step API that separates out the "mark as done" and "get next ready" steps, we'd have to either add API for mutating `Graph` or (better, IMO) add an extra field to store packages that are ready to be taken in a `take_ready` function.

Given the extra complication, I prefer the "keep it" option...

epage replied:
Do we need to mutate the graph or only `weights`?  If we did so, then wouldn't it still be trivial to track what packages are ready (its `self.weights.keys()`).

jneem replied:
Oh sorry, for some reason I misunderstood that you were talking about the `new` function. Yes, this can have a nicer API.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -498,3 +564,152 @@ fn transmit(
+    }
+}
+
+fn validate_registry(pkgs: &[&Package], reg_or_index: Option<&RegistryOrIndex>) -> CargoResult<()> {
```
nit: could extracting this be put in its own refactor commit?

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+            } else {
+                DEFAULT_TIMEOUT
+            };
+            if 0 < timeout {
+                let timeout = Duration::from_secs(timeout);
+                wait_for_publish(opts.gctx, source_ids.original, &to_confirm, timeout)?
+            } else {
+                Vec::new()
```
The timeout behavior is not intuitive and should at least be called out in the PR description and, once merged, in the tracking issue so we make sure we evaluate this when determining whether its ready for stabilization

epage commented on tests/testsuite/publish.rs:
```diff
@@ -961,7 +961,6 @@ fn publish_failed_with_index_and_only_allowed_registry() {
         .arg(registry.index_url().as_str())
         .with_status(101)
         .with_stderr_data(str![[r#"
-[NOTE] found `alternative` as only allowed registry. Publishing to it automatically.
```
Why was the note removed?

jneem replied:
`infer_registry` doesn't print this warning if `--index` or `--registry` was provided. The warning is misleading: we didn't automatically choose where to publish.

epage replied:
Thanks for that fix!

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
-        )?));
-    }
+    let (mut order, mut ready) = PublishOrder::new(&pkg_dep_graph.graph);
+    let mut to_confirm: HashSet<_> = ready.iter().copied().collect();
```
For readability, it feels weird that `to_confirm` has things that are `ready` but not published.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,151 +64,192 @@ pub struct PublishOpts<'gctx> {
+            };
+            if 0 < timeout {
+                let timeout = Duration::from_secs(timeout);
+                wait_for_publish(opts.gctx, source_ids.original, &to_confirm, timeout)?
```
Publish and waiting order is non-deterministic.  `ready` is created from a `HashMap` and `to_confirm` is a `HashSet`.

Internally, this makes testing a mess (likely the tests will break if `.unordered()` isn't used).

For users, its generally nice to put things in a deterministic order.

For `wait_for_publish`, we should probably wait in the order of upload to reduce the risk of hitting the timeout unnecessarily.

jneem replied:
I'll sort both `ready` and `to_confirm` so they'll be in the same order. In case it wasn't clear, btw, `wait_for_publish` polls all the packages on each run through the loop.

epage commented on tests/testsuite/publish.rs:
```diff
@@ -3758,7 +3846,23 @@ You may press ctrl-c to skip waiting; the crate should be available shortly.
         .with_status(101)
         .with_stderr_data(str![[r#"
-[ERROR] the `-p` argument must be specified to select a single package to publish
+[UPDATING] crates.io index
+[PACKAGING] a v0.0.1 ([ROOT]/foo/a)
+[PACKAGED] 3 files, [FILE_SIZE]B ([FILE_SIZE]B compressed)
+[PACKAGING] b v0.0.1 ([ROOT]/foo/b)
+[PACKAGED] 3 files, [FILE_SIZE]B ([FILE_SIZE]B compressed)
+[VERIFYING] a v0.0.1 ([ROOT]/foo/a)
+[COMPILING] a v0.0.1 ([ROOT]/foo/target/package/a-0.0.1)
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+[VERIFYING] b v0.0.1 ([ROOT]/foo/b)
+[UPDATING] crates.io index
+[ERROR] failed to verify package tarball
+
+Caused by:
+  failed to get `a` as a dependency of package `b v0.0.1 ([ROOT]/foo/target/package/b-0.0.1)`
+
+Caused by:
+  found a package in the remote registry and the local overlay: a@0.0.1
```
I'm assuming #14448 will improve this.  its a race for who is merged first and causes the other a conflict :)

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,187 +65,224 @@ pub struct PublishOpts<'gctx> {
+            // If nothing finished, it means we timed out while waiting for confirmation.
+            // We're going to exit, but first we need to check: if we pretend that the unconfirmed
+            // uploads are confirmed, did we finish everything?
+            for id in &ready {
+                remaining.remove(id);
+            }
```
Would it work to do this?  I feel like it would make the intent more clear.
```suggestion
            for id in &to_confirm {
                remaining.remove(id);
            }
```

jneem replied:
Not only is the intent more clear, the `ready` one was incorrect in some cases. I tried adding a test to exercise this, but it relies on specific timing that the test harness doesn't offer (we'd have to upload two crates at once, but only see one as ready when we poll).

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,187 +65,224 @@ pub struct PublishOpts<'gctx> {
+            }
+        }
+
+        let finished = if opts.dry_run {
```
Since we use the term `to_confirm`, it seems like it would make it easier to follow if we made this match
```suggestion
        let confirmed = if opts.dry_run {
```

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,187 +65,224 @@ pub struct PublishOpts<'gctx> {
+        }
+
+        let finished = if opts.dry_run {
+            ready.clone()
```
If we wait on `to_confirm`, why don't we return that here?
```suggestion
            to_confirm.clone()
```

jneem replied:
Sure, they should be the same in this case (except that one is a Vec and one is a BTreeSet; I've made them both BTreeSets)

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -57,187 +65,224 @@ pub struct PublishOpts<'gctx> {
+                break;
+            } else {
+                let failed_list = package_list(remaining, "and");
+                bail!("failed to publish {failed_list} because we timed out waiting for dependencies.");
```
Feels weird to say something failed that we haven't attempted
```suggestion
                bail!("unable to publish {failed_list} because we timed out waiting for dependencies.");
```

jneem commented:
Ok, I think I've addressed all of the comments. I tweaked the main upload loop to be hopefully a bit clearer (we're checking `remaining.is_empty()` instead of `to_confirm.is_empty()` for termination, and as a result we get to populate `to_confirm` after the upload instead of before).

I've called out the behavioral questions (the behavior of unpublishable packages, and the behavior of timeouts) in the PR description. I don't *know* of any other design questions that would need an unstable flag. I was thinking that the main application of packaging workspaces is for publishing workspaces, so there isn't much value in stabilizing one before the other.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -485,6 +586,74 @@ fn transmit(
+        let mut ready: Vec<_> = self
+            .weights
+            .iter()
+            .filter_map(|(id, weight)| (*weight == 0).then_some(*id))
```
There a reason we do this rather than remove them in `mark_published`?

This is sort of a continuation of https://github.com/rust-lang/cargo/pull/14433#pullrequestreview-2273538538

jneem replied:
Graph nodes aren't removable (and with the current graph representation, making them removable has possible footguns: either we allow dangling edges to nodes that no longer exist, or we need to scan the whole graph on each node removal, making it easy to accidentally introduce quadratic behavior).

Rather than mess with the graph type, I've added some additional state to `PublishOrder` that hopefully makes things more clear.

jneem replied:
Ah, thanks for the clarification. Yes, I think this is nicer and the names look good to me. My only concern is that I think this makes the whole process quadratic in the number of packages, since `take_ready` needs to scan over all the remaining packages. There's an intermediate solution that avoids the quadratic behavior by keeping the `ready` field (but still removing the `remaining` field). But also maybe workspaces are small enough to not worry about it...

epage replied:
Large workspaces are likely a hundred members or so.  Publish performance isn't in a critical path as well.

jneem replied:
Sounds good. I've pulled in your changes and squashed them in.

epage commented:
> I've called out the behavioral questions (the behavior of unpublishable packages, and the behavior of timeouts) in the PR description. 

You had:

> What is the desired behavior for the publish timeout? This PR uploads the crates in batches (depending on the dependency graph), and it applies the timeout for each batch. So for example, if you have a dependency chain `a -> b -> c` then we'll wait up to 60 seconds for `a`, and then up to 60 seconds for `b`, and then up to 60 seconds for `c`.

But I don't think a straight dependency tree was the concern.  Is my update accurate?

> What is the desired behavior for the publish timeout? This PR uploads the crates in batches (depending on the dependency graph), and we only timeout if nothing in the batch is available within the timeout, deferring the rest to the next wait-for-publish. So for example, if you have packages `a`, `b`, `c` then we'll wait up to 60 seconds and if only `a` and `b` were ready in that time, we'll then wait another 60 seconds for `c`.

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -115,82 +123,165 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+    // Then `to_confirm` will contain the new packages we just uploaded, plus
+    // most of the packages from the previous round.
+    let mut order = PublishOrder::new(&pkg_dep_graph.graph);
+    let mut ready = order.ready();
+    ready.sort();
```
`ready` returns sorted data

epage replied:
Could we just make that an invariant with a `BTreeSet`?

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -115,82 +123,165 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+            }
+        }
+
+        to_confirm.extend(ready.iter().copied());
```
When reading this, it implies to me that `ready` will continue to be used.

Some ideas
- `to_confirm.extend(ready.drain())` (depending on `ready`s type)
- `to_confirm.extend(std::mem::take(&mut ready)` 
- Inserting into `to_confirm` on the line right after `transmit`

... wait, does `ready` need to cross iterations of the loop?  We could loop over `remaining` and then immediately create `ready` before looping over it.  I feel like that would be a big help for understanding the code.

very minor nit: If we don't go with the last option, we *might* want to move `to_confirm` up a line at least.  Blank lines in code convey logical grouping.  Seeing a line like this suggests to me that we are initializing `to_confirm` for the following line but its for that and for future iterations of the loop.  Attaching it to the transmit operation (directly after it or at least after the block without a newline) conveys that what we transmit must be confirmed at some point.

jneem replied:
I went with the last option.

I also pushed some of the state into `PublishOrder` so that the loop is hopefully clearer. It would be also be possible to remove some of that state altogether, but (AFAICS) only at the cost of making things quadratic in the number of packages. (For example, we could remove ready, and just recalculate it every time as the set of remaining packages with weight zero.) Maybe being quadratic in the number of packages isn't important? But on the other hand I think the state is reasonably well-encapsulated now...

epage commented on src/cargo/ops/registry/publish.rs:
```diff
@@ -115,82 +123,165 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+
+        for id in &confirmed {
+            to_confirm.remove(id);
+            remaining.remove(id);
```
What if we removed from `remaining` when we transmit?  The loop will need to be updated to be `while !remaining.is_empty() || !to_confirm.is_empty()` which might negate the need for the comment I suggested.

Wait, do we need `remaining`?  If `PublishOrder` removes 0 weights, we can have a `any_ready` function that checks if weights is 0.  By removing `remaining`, we reduce the number of data structure we need to keep in sync and remove the question of "remaining to do what" (I keep assuming "to upload" while currently its" to do anything").

jneem commented:
> But I don't think a straight dependency tree was the concern. Is my update accurate?

Yes, your update is accurate. I thought that a long straight dependency tree might also be a concern, because it effectively extends the timeout.

epage commented:
> Yes, your update is accurate. I thought that a long straight dependency tree might also be a concern, because it effectively extends the timeout.

Not concerned about the straight line because the timeout is for publishing a package, not publishing everything, so its natural they get concatenated when doing them back-to-back.

epage commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit a016e5f5c2fab4edaa504e4928ca6bd747db0fbe has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage a016e5f5c2fab4edaa504e4928ca6bd747db0fbe -->
<!-- homu: {"type":"Approved","sha":"a016e5f5c2fab4edaa504e4928ca6bd747db0fbe","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit a016e5f5c2fab4edaa504e4928ca6bd747db0fbe with merge be1bbda84b39908f499a8826c4682da70d625422...
<!-- homu: {"type":"BuildStarted","head_sha":"a016e5f5c2fab4edaa504e4928ca6bd747db0fbe","merge_sha":"be1bbda84b39908f499a8826c4682da70d625422"} -->

epage commented:
btw I've noted in #10948 the open questions

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/10740753723/job/29790925207)
Approved by: epage
Pushing be1bbda84b39908f499a8826c4682da70d625422 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/10740753723/job/29790925207"},"merge_sha":"be1bbda84b39908f499a8826c4682da70d625422"} -->
